src/nti/webhooks/delivery_manager.py

```diff
@@ -28,14 +35,52 @@ def __init__(self, subscriptions_and_attempts):
         self._sub_and_attempts = list(subscriptions_and_attempts)
 
     def deliver(self):
+        # pylint:disable=broad-except
         for sub, attempt in self._sub_and_attempts:
-            response = requests.post(sub.to, data=attempt.payload_data)
+            attempt.request.createdTime = time.time()
+            try:
+                response = requests.post(sub.to, data=attempt.payload_data)
+            except Exception as ex:
+                logger.exception("Failed to deliver for attempt %s/%s", sub, attempt)
+                attempt.response = None
+                attempt.status = 'failed'
+                attempt.message = str(ex)
+                continue
+
             attempt.status = 'successful' if response.ok else 'failed'
-            # XXX: Store the full response status, data (to some limit) and headers.
-            # We don't want to pickle the Response object to avoid depending on internal
-            # details, we need to extract it.
             attempt.message = u'%s %s' % (response.status_code, response.reason)
 
+            try:
+                self._fill_req_resp_from_request(attempt, response)
+                # This is generally a programming error, probably an encoding something
+                # or other.
+            except Exception as ex:
+                logger.exception("Failed to parse response for attempt %s/%s", sub, attempt)
+                attempt.message = str(ex)
+
+    if str is bytes:
+        def _dict_to_text(self, headers):
+            return {text_type(k): text_type(v) for k, v in headers.iteritems()}
+    else:
+        _dict_to_text = dict
+
+    def _fill_req_resp_from_request(self, attempt, http_response):
+        http_request = http_response.request # type: requests.PreparedRequest
+        req = attempt.request
+        rsp = attempt.response
+        rsp.createdTime = time.time()
+
+        req.url = http_request.url
+        req.method = text_type(http_request.method)
+        req.body = text_type(http_request.body) # XXX: Text/bytes. This uses default encoding.
+        req.headers = self._dict_to_text(http_request.headers)
```

Comment:
Do we need to pop off any headers that could have secrets in them? `Set-Cookie`, etc?

Reply:
Possibly. I have a comment about that in the interface definition. But as I was implementing, I couldn't come up with a scenario where that actually matters (currently). Outgoing hook deliveries include no authentication, so there's nothing on that side, and no reason to expect that a response will include anything either. 

I looked at hooks configured for several different services and didn't find anything that needed to be dropped. 

We'll know more as we make real-life deliveries.